pilot/pkg/networking/core/v1alpha3/httproute.go

```diff
@@ -231,10 +232,38 @@ func TranslateRoutes(in model.Config, name ClusterNaming) []GuardedRoute {
 // TranslateRoute translates HTTP routes
 // TODO: fault filters -- issue https://github.com/istio/api/issues/388
 func TranslateRoute(in *networking.HTTPRoute,
-	match *networking.HTTPMatchRequest,
+	match *networking.HTTPMatchRequest, port int,
 	operation string,
-	name ClusterNaming) GuardedRoute {
-	out := route.Route{
+	name ClusterNaming,
+	proxyLabels model.LabelsCollection,
+	gatewayName string) *route.Route {
+
+	sourceMatched := false
+	// Trim by source labels or mesh gateway
+	if match != nil {
+		if len(match.Gateways) > 0 {
+			for _, g := range match.Gateways {
+				if g == gatewayName {
+					sourceMatched = true
+					break
+				}
+			}
+		} else if proxyLabels.IsSupersetOf(match.GetSourceLabels()) {
+			sourceMatched = true
+		}
+	} else {
+		sourceMatched = true
+	}
+
+	if !sourceMatched {
+		return nil
+	}
+
+	if match != nil && match.Port != nil && match.Port.GetNumber() != uint32(port) {
```

Comment:
Move this and the above logic to a function so we can just do
```go
if !SourceMatch(...) {
   return nil
}
```
and keep the function very clean.